File: yumi/sdk/kotlin/src/main/kotlin/io/yumi/sdk/Connection.kt
package io.yumi.sdk

data class ConnectionConfig(
    val mode: String,
    val baseUrl: String,
    val accessToken: String,
) {
    fun websocketUrl(): String =
        if (mode == "relay") "${httpToWs(baseUrl.trimEnd('/'))}/ws/edge" else baseUrl
}

fun httpToWs(url: String): String {
    val u = url.trimEnd('/')
    return when {
        u.startsWith("https://") -> "wss://" + u.substring("https://".length)
        u.startsWith("http://") -> "ws://" + u.substring("http://".length)
        else -> u
    }
}

fun resolveConnection(code: String, edgeName: String): ConnectionConfig {
    val relayUrl = yumiEnv("YUMI_RELAY_URL")
    val accessToken = yumiEnv("YUMI_ACCESS_TOKEN")
    if (relayUrl.isNotEmpty() && accessToken.isNotEmpty()) {
        return ConnectionConfig("relay", relayUrl.trimEnd('/'), accessToken)
    }

    if (code.startsWith("ws://") || code.startsWith("wss://")) {
        return ConnectionConfig("direct", code, "")
    }

    if (isLanCode(code)) {
        val serverUrl = parseLanCodeToServerUrl(code)
        val wsUrl = "${httpToWs(serverUrl)}/ws/edge"
        return ConnectionConfig("direct", wsUrl, "")
    }

    if (isRelayToken(code)) {
        val profile = bootstrapProfile(code, "edge", edgeName)
        setYumiEnv("YUMI_RELAY_URL", profile.relayUrl)
        setYumiEnv("YUMI_ACCESS_TOKEN", profile.accessToken)
        return ConnectionConfig("relay", profile.relayUrl, profile.accessToken)
    }

    if (code.startsWith("http://") || code.startsWith("https://")) {
        val base = code.trimEnd('/')
        val wsUrl = "${httpToWs(base)}/ws/edge"
        return ConnectionConfig("direct", wsUrl, "")
    }

    return ConnectionConfig("direct", "ws://127.0.0.1:8000/ws/edge", "")
}

File: yumi/sdk/go/connection.go
package yumi_sdk

import (
	"os"
	"strings"
)

// ConnectionConfig holds resolved connection details.
type ConnectionConfig struct {
	Mode        string // "direct" or "relay"
	BaseURL     string
	AccessToken string
}

func (c *ConnectionConfig) relayEdgeWsURL() string {
	return httpToWs(strings.TrimRight(c.BaseURL, "/")) + "/ws/edge"
}

func httpToWs(u string) string {
	if strings.HasPrefix(u, "https://") {
		return "wss://" + u[len("https://"):]
	}
	if strings.HasPrefix(u, "http://") {
		return "ws://" + u[len("http://"):]
	}
	return u
}

// ResolveConnection determines the WebSocket connection target from
// environment variables and the user-provided connection code.
func ResolveConnection(code, edgeName string) (*ConnectionConfig, error) {
	relayURL := os.Getenv("YUMI_RELAY_URL")
	accessToken := os.Getenv("YUMI_ACCESS_TOKEN")
	if relayURL != "" && accessToken != "" {
		return &ConnectionConfig{
			Mode:        "relay",
			BaseURL:     strings.TrimRight(relayURL, "/"),
			AccessToken: accessToken,
		}, nil
	}

	if strings.HasPrefix(code, "ws://") || strings.HasPrefix(code, "wss://") {
		return &ConnectionConfig{Mode: "direct", BaseURL: code}, nil
	}

	if isLanCode(code) {
		serverURL, err := parseLanCode(code)
		if err != nil {
			return nil, err
		}
		wsURL := httpToWs(strings.TrimRight(serverURL, "/")) + "/ws/edge"
		return &ConnectionConfig{Mode: "direct", BaseURL: wsURL}, nil
	}

	if isRelayToken(code) {
		profile, err := BootstrapProfile(code, "edge", edgeName)
		if err != nil {
			return nil, err
		}
		os.Setenv("YUMI_RELAY_URL", profile.RelayURL)
		os.Setenv("YUMI_ACCESS_TOKEN", profile.AccessToken)
		return &ConnectionConfig{
			Mode:        "relay",
			BaseURL:     profile.RelayURL,
			AccessToken: profile.AccessToken,
		}, nil
	}

	if strings.HasPrefix(code, "http://") || strings.HasPrefix(code, "https://") {
		wsURL := httpToWs(strings.TrimRight(code, "/")) + "/ws/edge"
		return &ConnectionConfig{Mode: "direct", BaseURL: wsURL}, nil
	}

	return &ConnectionConfig{
		Mode:    "direct",
		BaseURL: "ws://127.0.0.1:8000/ws/edge",
	}, nil
}
